### PR TITLE
fix: Multiple QoL fixes. See #1017 desc

### PR DESCRIPTION
**Pull request by Ambryal**
- Set clean shaven beard type avalaible
- Set CellLogs Clickable for
                                         - New student
                                         - Student levelUp
- April fools display 
                           • for the bosses
                                         - Emberfang
                                         - GoldClaw
                           • in
                                         - Boss texture
                                         - Boss HUD name
                                         - Boss name in ClanView
- Cerberus slight number fixes
- Multibrute fights now dynamically display correct
                                                         - Bust image
                                                         - Name
                                                         - Weapons
                                                         - Tooltip
- Hammer now properly display net breaking in [this fight](https://brute.eternaltwin.org/Gargamelpourchat/fight/c5e5bbf4-aa31-48cc-bd67-a110f0df067a) 
- Item drop positions fix
- Drop shield before using Hammer and Flashflood
- Hammer disarms weapon and shield
- Leveling up send you to levelUp page if you can levelUp again
- Link to cell added in levelUp page

### CODE REVIEW

File: client/src/assets/i18n/en.json [hunk 1]
```diff
@@ -572,6 +572,7 @@
   "fight.step.stun": "{{brute}} was stunned.",
   "fight.step.haste": "{{brute}} rushed past {{opponent}} and dealt {{damage}} damage.",
   "fight.step.treat": "{{brute}} gave their pet {{target}} a treat and healed them for {{heal}}HP.",
+  "fight.step.dropShield": "{{brute}} dropped it's shield.",
```
Comment (Zenoo) [inline client/src/assets/i18n/en.json]: Using neutral pronouns is a better choice whenever possible

```suggestion
  "fight.step.dropShield": "{{brute}} dropped their shield.",
```

File: client/src/utils/fight/setupFight.ts [hunk 2]
```diff
@@ -386,60 +381,79 @@ const setupFight: (
...
+                ? bossPhantomHpBar
+                : undefined) ?? undefined,
+        bustImage: (fighter.master
+          ? undefined
+          : fighter.id === brute1.id
+            ? brute1BustImg
+            : fighter.id === brute2?.id
+              ? brute2BustImg
+              : fighter.type === 'brute'
+                ? await createBustImage(fighter, renderer)
```
Comment (Zenoo) [inline client/src/utils/fight/setupFight.ts]: This whole .map has been asynced only for this.
It will impact performance on mobile when more than 2 brutes are rendered.
Instead, revert the .map to be non asynced, and only create the 2 first busts before the .map, and generate the others only when needed.

File: client/src/utils/fight/updateWeapons.ts [hunk 3]
```diff
@@ -20,13 +20,13 @@ const updateWeapons = (
...
-  // Only affect the UI for main brutes
-  if (brute.bust) {
+  // Only affect the HUD for the focused brutes
+  if (brute.HUDFocused) {
     // Empty list
     brute.weaponsIllustrations.forEach((illustration) => {
       illustration.destroy();
     });
-    brute.weaponsIllustrations = [];
+    brute.weaponsIllustrations.splice(0);
```
Comment (Zenoo) [inline client/src/utils/fight/updateWeapons.ts]: Why this change?

splice returns the spliced elements, meaning more memory assigned, where simply assigning a new array doesn't.
As the illustration references are destroyed above, assigning a new array does not create a memory leak.

File: client/src/utils/fight/utils/findFighter.ts [hunk 4]
```diff
@@ -17,4 +20,13 @@ const findFighter = (
   stepFighter?: number,
 ) => (typeof stepFighter === 'undefined' ? undefined : fighters.find((f) => f.index === stepFighter));
 
+export const findHUDFocusedFighter = (
+  fightFighters: Fighter[],
+  fighters: AnimationFighter[],
+  team: string,
+) => {
+  const animationFighter = fighters.find((f) => f.HUDFocused && f.team === team);
+  return fightFighters?.find((f) => f.id === animationFighter?.id);
```
Comment (Zenoo) [inline client/src/utils/fight/utils/findFighter.ts]: fightFighters is not declared optional above
```suggestion
  return fightFighters.find((f) => f.id === animationFighter.id);
```

File: server/src/utils/fight/fightMethods.ts [hunk 5]
```diff
@@ -1355,6 +1388,18 @@ const breakShield = (fighter: DetailedFighter, opponent: DetailedFighter) => {
   return getFighterStat(fighter, 'disarm') * 100 >= randomBetween(1, 300);
 };
 
+const dropShield = (fightData: DetailedFight, fighter: DetailedFighter) => {
+  // Remove brute's shield
+  fighter.shield = false;
+  fighter.block -= SHIELD_BLOCK_ODDS;
```
Comment (Zenoo) [inline server/src/utils/fight/fightMethods.ts]: The way to handle shield drops have changed since your modifications, please update it. (skills array update + no more odds constant)

**Review by Zenoo — CHANGES_REQUESTED** [review]
Translation files other than `en.json` should never be edited, please revert your changes on them.
They are handled by Crowdin.

fightMethod has also changed since your modifications, please rebase your branch.

Other than that, I noted a few things to change, but everything else looks good, thanks !

It would be better for future contributions to list the changes in the PR description, for easier retracing if needed

File: client/src/utils/fight/updateWeapons.ts (hunk 3 above)
Comment (Ambryal) [reply client/src/utils/fight/updateWeapons.ts]: This change was made because the weaponIllustration array is shared with every brute. Should I 
1) Rename it "teamWeaponIllustrations" ? 
2) Or call a new method "destroyWeapons" for previously displayed brute, then updateWeapons for newly displayed one ?

Comment (Zenoo) [reply client/src/utils/fight/updateWeapons.ts]: I don't understand your logic.
splice(0) does the same thing as  = [], as it empties the array.

It doesn't change anything other than waste memory.

Comment (Ambryal) [reply client/src/utils/fight/updateWeapons.ts]: = [] creates a new empty array, whereas splice(0) empty the array, this is not the same thing

Here, I need brute.weaponsIllustrations array to keep it's pointer as this array is in fact the team weaponIllustration array, which is emptied and filled everytime a new brute is displayed.

Therefore, either you are ok with just renaming it "teamWeaponIllustration" for clarity, or I make weaponIllustration not shared anymore but I would have to add a new "removeWeaponsHUD" function or add a "previousDisplayedBrute" agrument to updateWeapons.

Comment (Ambryal) [reply client/src/utils/fight/updateWeapons.ts]: I could also do it like this :

while (brute.weaponsIllustrations.length){

  brute.weaponsIllustrations.pop().destroy();
  
}

But I don't know how ok you are with while loops

Comment (Zenoo) [reply client/src/utils/fight/updateWeapons.ts]: I didn't realize everyone on the team shared the same array reference.
In this case, renaming it is better.
Also, don't use splice, use .length = 0 to clear it instead.

File: client/src/utils/fight/setupFight.ts (hunk 2 above)
Comment (Ambryal) [reply client/src/utils/fight/setupFight.ts]: This would require to add the   renderer: RendererContextInterface, as argument for every fight step that can change displayed brutes.

Do you want me to do this or do you prefer if instead of my current approach of adding setHUDFocus to every step, I just check if fightStep have "t,b, or f" in the main loop, for less "encombrement" ?

Comment (Ambryal) [reply client/src/utils/fight/setupFight.ts]: Opted for a version of setHUDFocus that looked easier to read and to maintain.

File: client/src/utils/fight/setupFight.ts [hunk 6]
```diff
@@ -464,6 +475,19 @@ const setupFight: (
...
+      void setHUDFocus(app, renderer, fighters, step.b, speed, isClanWar);
+    }
+    // Display step's fighter in HUD
+    if ('f' in step) {
+      void setHUDFocus(app, renderer, fighters, step.f, speed, isClanWar);
+    }
+    // Display step's target in HUD
+    if ('t' in step && step.a !== StepType.Bomb) {
+      void setHUDFocus(app, renderer, fighters, step.t, speed, isClanWar);
+    }
```
Comment (Zenoo) [inline client/src/utils/fight/setupFight.ts]: Using `Object.hasOwn` is safer, as it has less edge cases

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn

File: server/src/utils/fight/fightMethods.ts [hunk 7]
```diff
@@ -1374,6 +1407,19 @@ const breakShield = (fighter: DetailedFighter, opponent: DetailedFighter) => {
   return getFighterStat(fighter, 'disarm') * 100 >= randomBetween(1, 300);
 };
 
+const dropShield = (fightData: DetailedFight, fighter: DetailedFighter) => {
+  // Remove brute's shield
+  fighter.shield = false;
+  fighter.skills = fighter.skills.filter((sk) => sk.name !== SkillName.shield);
+  fighter.block -= SkillModifiers[SkillName.shield][FightStat.BLOCK]?.percent ?? 0;
```
Comment (Zenoo) [inline server/src/utils/fight/fightMethods.ts]: I made a typo on the way I handle block stat changes, it should now be
```suggestion
  fighter.block -= (SkillModifiers[SkillName.shield][FightStat.BLOCK]?.percent ?? 0) / 100;
```

**Review by Zenoo — CHANGES_REQUESTED** [review]

File: client/src/utils/fight/setupFight.ts (hunk 6 above)
Comment (Ambryal) [reply client/src/utils/fight/setupFight.ts]: I had to keep the "in" to get eslinted though

File: client/src/views/LevelUpView.tsx [hunk 8]
```diff
@@ -30,6 +31,11 @@ const LevelUpView = () => {
   const { brute, updateBrute } = useBrute();
   const theme = useTheme();
 
+  const xpNeededForNextLevel = useMemo(
+    () => (brute ? getXPNeeded(brute.level + 1) : 0),
+    [brute],
+  );
```
Comment (Zenoo) [inline client/src/views/LevelUpView.tsx]: No need to memo this, as it's only used once in a callback.
It doesn't need to be declared

File: client/src/views/LevelUpView.tsx [hunk 9]
```diff
@@ -74,8 +80,8 @@ const LevelUpView = () => {
     // Update brute data
     updateBrute((data) => (data ? ({ ...data, ...newBrute }) : null));
 
-    navigate(`/${brute.name}/cell`);
-  }, [Alert, brute, choices, navigate, updateBrute, updateData]);
+    navigate(xpNeededForNextLevel > brute.xp ? `/${brute.name}/cell` : `/${brute.name}/level-up`);
```
Comment (Zenoo) [inline client/src/views/LevelUpView.tsx]: Instead, just use `getXPNeeded(brute.level + 1)` directly here

**Review by Zenoo — CHANGES_REQUESTED** [review]

File: client/src/views/LevelUpView.tsx (hunk 9 above)
Comment (Ambryal) [reply client/src/views/LevelUpView.tsx]: Sorry i'll have to watch some tutorial

Comment (Ambryal) [reply client/src/views/LevelUpView.tsx]: I think I did something wrong with git it pushed some files I never added

**Review by Zenoo — APPROVED** [review]
Everything looks good code-wise, I will do some further testing before merging.
Could you list all the changes in the PR description please?

**Comment by Ambryal** [comment]
Yeah sorry, it's updated now

**Comment by Zenoo** [comment]
Everything seems to work.
Thanks for your contribution.
